File: backend/app/schemas/everyday_safety.py
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.audit import AuditSummary
from app.schemas.recalls import RecallRiskScore


EverydaySafetyCategory = Literal["food_supplement"]
EverydaySafetySourceType = Literal[
    "FDA_FOOD_ENFORCEMENT",
    "USDA_FSIS_RECALL",
    "FDA_PUBLIC_NOTICE",
    "FDA_NORMALIZED_PUBLIC_NOTICE",
]
EverydaySafetySourceKind = Literal[
    "structured_api",
    "public_notice",
    "normalized_public_notice",
]


class EverydaySafetySource(BaseModel):
    name: str
    endpoint: str
    retrieval_timestamp: str
    source_kind: EverydaySafetySourceKind = "structured_api"
    source_type: str = "official API record"


class EverydaySafetyCheckedSource(BaseModel):
    source_id: str
    source_name: str
    source_type: EverydaySafetySourceType
    endpoint: str
    source_kind: EverydaySafetySourceKind = "structured_api"
    record_type: str = "official API record"
    upstream_status: str
    record_count: int


class EverydaySafetyRecord(BaseModel):
    record_id: str | None = None
    recall_number: str | None = None
    product_description: str | None = None
    reason_for_recall: str | None = None
    classification: str | None = None
    status: str | None = None
    recall_initiation_date: str | None = None
    report_date: str | None = None
    distribution_pattern: str | None = None
    recalling_firm: str | None = None
    product_quantity: str | None = None
    code_info: str | None = None
    source_type: EverydaySafetySourceType
    source_kind: EverydaySafetySourceKind = "structured_api"
    source_record_type: str = "official API record"
    title: str | None = None
    product_name: str | None = None
    brand_name: str | None = None
    company_name: str | None = None
    remedy: str | None = None
    official_url: str | None = None
    affected_models: list[str] = Field(default_factory=list)
    affected_lots: list[str] = Field(default_factory=list)
    extraction_confidence: str | None = None
    source_text_excerpt: str | None = None
    search_strategy_used: str
    risk_score: RecallRiskScore
    source: EverydaySafetySource


class EverydaySafetySearchResponse(BaseModel):
    query: str
    raw_query: str
    normalized_query: str
    correction_applied: bool
    suggestion_message: str | None = None
    category: EverydaySafetyCategory
    category_label: str
    count: int
    limit: int
    source_name: str
    endpoint: str
    retrieval_timestamp: str
    score_version: str
    search_strategy_used: str
    sources_checked: list[EverydaySafetyCheckedSource]
    public_data_disclaimer: str
    limitations: list[str]
    audit: AuditSummary
    results: list[EverydaySafetyRecord]

File: frontend/src/api/everydaySafety.ts
import { apiClient } from './client'

export type EverydaySafetySourceType =
  | 'FDA_FOOD_ENFORCEMENT'
  | 'USDA_FSIS_RECALL'
  | 'FDA_PUBLIC_NOTICE'
  | 'FDA_NORMALIZED_PUBLIC_NOTICE'

export type EverydaySafetySourceKind =
  | 'structured_api'
  | 'public_notice'
  | 'normalized_public_notice'

export type EverydaySafetyCategory = 'food_supplement'
export type EverydaySafetySort = 'score' | 'latest'

export type EverydaySafetyScoreComponents = {
  classification_score: number
  status_score: number
  recency_score: number
  scope_score: number
}

export type EverydaySafetyRiskScore = {
  score: number
  label: string
  components: EverydaySafetyScoreComponents
  score_version: string
}

export type EverydaySafetySource = {
  name: string
  endpoint: string
  retrieval_timestamp: string
  source_kind?: EverydaySafetySourceKind
  source_type?: string
}

export type EverydaySafetyCheckedSource = {
  source_id: string
  source_name: string
  source_type: EverydaySafetySourceType
  source_kind?: EverydaySafetySourceKind
  record_type?: string
  endpoint: string
  upstream_status: string
  record_count: number
}

export type EverydaySafetyRecord = {
  record_id: string | null
  recall_number: string | null
  product_description: string | null
  reason_for_recall: string | null
  classification: string | null
  status: string | null
  recall_initiation_date: string | null
  report_date: string | null
  distribution_pattern: string | null
  recalling_firm: string | null
  product_quantity: string | null
  code_info: string | null
  source_type: EverydaySafetySourceType
  source_kind?: EverydaySafetySourceKind
  source_record_type?: string
  title?: string | null
  product_name?: string | null
  brand_name?: string | null
  company_name?: string | null
  remedy?: string | null
  official_url?: string | null
  affected_models?: string[]
  affected_lots?: string[]
  extraction_confidence?: string | null
  source_text_excerpt?: string | null
  search_strategy_used: string
  risk_score: EverydaySafetyRiskScore
  source: EverydaySafetySource
}

export type EverydaySafetyAudit = {
  audit_id: string
  source_id: string
  module: string
  upstream_status: string
  record_count: number
  transform_version: string
  source_snapshot_status: string | null
  source_pull_id: string | null
  source_payload_hash: string | null
}

export type EverydaySafetySearchResponse = {
  query: string
  raw_query?: string
  normalized_query?: string
  correction_applied?: boolean
  suggestion_message?: string | null
  category: EverydaySafetyCategory
  category_label: string
  count: number
  limit: number
  source_name: string
  endpoint: string
  retrieval_timestamp: string
  score_version: string
  search_strategy_used: string
  sources_checked: EverydaySafetyCheckedSource[]
  public_data_disclaimer: string
  limitations: string[]
  audit: EverydaySafetyAudit
  results: EverydaySafetyRecord[]
}

export async function searchEverydaySafety(
  query: string,
  limit = 5,
  category: EverydaySafetyCategory = 'food_supplement',
  sort: EverydaySafetySort = 'score'
): Promise<EverydaySafetySearchResponse> {
  const response = await apiClient.get<EverydaySafetySearchResponse>(
    '/api/v1/everyday-safety/search',
    {
      params: {
        category,
        q: query,
        limit,
        sort,
      },
    }
  )

  return response.data
}
